### ISSUE TITLE
Include default values in storage metadata

### ISSUE THREAD

**Issue by bkchr** [paritytech/substrate#1256]
Currently these values are not exposed.

@cheme maybe an easy task for you :)

**Comment by cheme** [comment]
I wonder if we want to:
1) expose the expression for building the default values
2) expose the values resulting from calling this expression.

First case is pretty straight forward (but there is still he added spaces issue) : https://github.com/paritytech/substrate/compare/master...cheme:defaultval-metadata .
Second case need to alter more code (we instantiate a value and some code is not really designed for that), it also need a conversion to string somehow so probably fmt::Debug constraint which is not enforced and may also produce not so good result (fmt::Display seems not realistic as it would require lot of additional implementation eg pair, array and all).

**Comment by xlc** [comment]
Why it needs to be able to convert to string? All storage fields should be encodable and the hex data could be the string form. Otherwise it could be tricky for js implementation to parse the debug string format.

**Comment by cheme** [comment]
Thanks, I did not think about it, I like the idea. 
Yet it is not quickly doable (StorageFunctionMetadata for a field is in an array so we cannot make the type parametric over the different storage fields types, we need a dyn type but it does not seem possible with current definition of Encode trait). Splitting encode trait to a subtrait with only encode function exposed would solve that, but I feel like it is not really a good move.

**Comment by bkchr** [comment]
I think solution 1 is the best one. As some fields are generic you would not know in javascript how to decode them. So, exposing the expression is the correct way in my opinion.

**Comment by bkchr** [comment]
If we have the following:
```
trait Trait {
    type AccountId;
}

struct GenesisConfig<T: Trait> {
    account_id: T::AccountId
}
```

The javascript side does not know anything about the type, it just sees `T::AccountId`, but it will never know what the actual type is. So, decoding will not work. And even encoding will not work, because you does not know the type and can not encode it. (the metadata is generated statically).

**Comment by bkchr** [comment]
`Default::default()` is not the best information, but I think that for these fields the value is not that important at all. Depending on where we want to go with the metadata, we can revisit this in a later pull request.

**Comment by xlc** [comment]
JS side must know the type ahead of time. Those information are currently coded at https://github.com/polkadot-js/api/tree/master/packages/types/src

`AccountId` may be generic to the module, but it must be a concrete type for a particular blockchain （[polkadot.js hardcoded it to u256](https://github.com/polkadot-js/api/blob/e99f17c22c428095a340c22c57b50738763216f3/packages/types/src/AccountId.ts#L23)).

The JS lib can encode the data and make transaction, can decode the data to query information, I don't understand why it can't handle default data in metadata.

The data stored in metadata is not enough to decode the value, which is an issue (https://github.com/paritytech/substrate/issues/917), but have expression form doesn't really help. How do you handle `DEFAULT_MINIMUM_VALIDATOR_COUNT` at all?

**Comment by bkchr** [comment]
I know that the type is concrete for a particular implementation. I have written the first version of all the metadata exposing in substrate, so I know the current limitations very well.
Even when the javascript side hardcodes that `T::AccountId` is `u256`, we still don't know this add the generation of the metadata. The metadata is generated at compile time and there we also don't know the concrete type.
In the ticket you have linked, the last sentence gives you an outlook on when we maybe could tackle this issue. In the current form of the metadata generation, we just don't have any other way of solving it. Sorry!

**Comment by xlc** [comment]
Thanks for the clarification. I think be able to generate and expose a more descriptive metadata is an very important feature. I will do some more research to see if I can improve the situation.

**Comment by bkchr** [comment]
Should be fixed by #1264
